src/ComponentInstaller.php

```diff
@@ -353,133 +347,153 @@ public function onPostPackageUninstall(PackageEvent $event)
      *
-     * @param array<string,mixed> $extra
-     * @return array<string,mixed>
+     * @return ComposerExtraComponentInstallerArrayType|ComposerExtraComponentInstallerProjectArrayType
+     * @psalm-return (
+     *  $rootProject is true
+     *  ? ComposerExtraComponentInstallerProjectArrayType
+     *  : ComposerExtraComponentInstallerArrayType
+     * )
      */
-    private function getExtraMetadata(array $extra): array
+    private function getExtraMetadata(array $extra, bool $rootProject = false): array
     {
         if (isset($extra['laminas']) && is_array($extra['laminas'])) {
-            /** @var array<string,mixed> $laminasSpecificConfiguration */
-            $laminasSpecificConfiguration = $extra['laminas'];
-            return $laminasSpecificConfiguration;
+            return $this->filterComponentInstallerSpecificValuesFromComposerExtra($extra['laminas'], $rootProject);
         }
 
         /**
          * supports legacy "extra.zf" configuration
-         *
-         * @var array<string,mixed> $deprecatedConfiguration
          */
-        $deprecatedConfiguration = isset($extra['zf']) && is_array($extra['zf'])
-            ? $extra['zf']
-            : [];
+        if (isset($extra['zf']) && is_array($extra['zf'])) {
+            return $this->filterComponentInstallerSpecificValuesFromComposerExtra($extra['zf'], $rootProject);
+        }
 
-        return $deprecatedConfiguration;
+        return [];
     }
 
     /**
      * Discover what package types are relevant based on what the package
      * exposes in the extra configuration.
      *
-     * @param array<string,mixed> $extra
-     * @return Collection Collection of Injector\InjectorInterface::TYPE_* constants.
+     * @param ComposerExtraComponentInstallerArrayType $extra
+     * @return Collection<non-empty-string,Injector\InjectorInterface::TYPE_*>
      */
     private function discoverPackageTypes(array $extra): Collection
     {
-        $packageTypes = array_flip($this->packageTypes);
-        $knownTypes   = array_keys($packageTypes);
-
-        return Collection::create($extra)
-            ->filter(function ($packages, $type) use ($knownTypes) {
-                return in_array($type, $knownTypes, true);
-            })
-            ->reduce(function (Collection $discoveredTypes, $packages, string $type) use ($packageTypes): Collection {
-                $packages = is_array($packages) ? $packages : [$packages];
-
-                foreach ($packages as $package) {
-                    assert(is_string($package));
-                    $discoveredTypes[$package] = $packageTypes[$type];
+        $packageTypes = array_flip(self::PACKAGE_TYPES);
+
+        /** @var Collection<non-empty-string,Injector\InjectorInterface::TYPE_*> $discoveredTypes */
+        $discoveredTypes = new Collection([]);
+        (new Collection($extra))
+            ->filter(
+            /** @psalm-suppress UnusedClosureParam https://github.com/vimeo/psalm/issues/8095 */
```

Comment:
It is possible to declare unused parameters with the `$_` or `$_unused` syntax. Just noting it here.

Reply:
It is, but AFAIR our codestyle will bitch due to the usage of `_`.

Reply:
Yup:

```shell
FILE: src/Injector/ConfigInjectorChain.php
------------------------------------------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
------------------------------------------------------------------------------------------------------------------------------------------
 55 | ERROR | Variable "$_" is not in valid camel caps format
    |       | (WebimpressCodingStandard.NamingConventions.ValidVariableName.NotCamelCaps)
------------------------------------------------------------------------------------------------------------------------------------------

```